tv_apps/tv-deck/js/channel_manager.js

```diff
@@ -25,7 +30,305 @@
+  var proto = Object.create(new evt());
+
+  // Current tuner object playing by TV (we may have another tuner for recording
+  // in the future).
+  Object.defineProperty(proto, 'playingTuner', {
+    get: function() {
+      return this.currentTuners[this.playingTunerId];
+    }
+  });
+
+  // Current source object playing by TV (we may have another source for
+  // recording in the future).
+  Object.defineProperty(proto, 'playingSource', {
+    get: function() {
+      if (!this.playingTuner || !this.playingTuner.sources) {
+        return null;
+      }
+      return this.playingTuner.sources[this.playingSourceType];
+    }
+  });
+
+  // Current channel object playing by TV (we may have another source for
+  // recording in the future).
+  Object.defineProperty(proto, 'playingChannel', {
+    get: function() {
+      var playingSource = this.playingSource;
+      var index = playingSource.channelIndexHash[this.playingChannelNumber];
+      if (!this.playingSource || (index !== 0 && !index)) {
+        return null;
+      }
+      return playingSource.channels[index];
+    }
+  });
+
+  // Current hash
+  Object.defineProperty(proto, 'currentHash', {
+    get: function() {
+      return this.playingTunerId + ',' + this.playingSourceType +
+             ',' + this.playingChannelNumber;
+    }
+  });
+
+  /**
+   * Fetch tuner, source and channel settings from URL hash. If URL hash is
+   * empty, then fetch settings from localStorage.
+   */
+  proto.fetchSettingFromHash = function cm_fetchSettingFromHash() {
+    return new Promise(function(resolve, reject) {
+      var hash;
+      var fetch = function() {
+        if(hash) {
+          hash = hash.split(',');
+          this.playingTunerId = hash[0];
+          this.playingSourceType = hash[1];
+          this.playingChannelNumber = hash[2];
+        }
+      }.bind(this);
+
+      hash = window.location.hash.substring(1);
+      if (hash.length === 0) {
+        asyncStorage.getItem('TV_Hash', function(item) {
+          hash = item;
+          fetch();
+          resolve();
+        });
+        return;
+      }
+
+      fetch();
+      resolve();
+    }.bind(this));
+  };
+
+  /**
+   * Retrieve all the currently available TV tuners.
+   */
+  proto.scanTuners = function cm_scanTuners() {
+    this.isScanning = true;
+    navigator.tv.getTuners().then(function onsuccess(tuners) {
+      this.isScanning = false;
+      this.currentTuners = {};
+      tuners.forEach(function initTuner(tuner) {
+        this.currentTuners[tuner.id] = {
+          tuner: tuner,
+          sources: {}
+        };
+      }.bind(this));
+
+      if (!this.playingTunerId) {
+        this.playingTunerId = tuners[0].id;
+      }
+
+      if (!this.playingTuner) {
+        this.fire('error');
+        return;
+      }
+
+      this.scanSources();
+    }.bind(this));
+  };
+
+  /**
+   * Retrieve all the currently available TV sources from current TV tuner.
+   */
+  proto.scanSources = function cm_scanSources() {
```

Comment:
I would suggest rename this function as _scanSources since it should not be a public function and only be called at scanTunner